src/string.cr

```diff
@@ -746,47 +746,82 @@ class String
     char_at(index) { raise IndexError.new }
   end
 
-  # Returns a substring by using a Range's *begin* and *end*
-  # as character indices. Indices can be negative to start
-  # counting from the end of the string.
+  # Returns the substring indicated by *range* as span of character indices.
   #
-  # Raises `IndexError` if the range's start is out of bounds.
+  # The substring ranges from `self[range.begin]` to `self[range.end]`
+  # (or `self[range.end - 1]` if the range is exclusive). It can be smaller than
+  # `range.size` if the end index is larger than `self.size`.
   #
   # ```
-  # "hello"[0..2]   # => "hel"
-  # "hello"[0...2]  # => "he"
-  # "hello"[1..-1]  # => "ello"
-  # "hello"[1...-1] # => "ell"
-  # "hello"[6..7]   # raises IndexError
+  # s = "abcde"
+  # s[1..3] # => "bcd"
+  # # range.end > s.size
+  # s[3..7] # => "de"
   # ```
-  def [](range : Range)
+  #
+  # Open ended ranges are clamped at the start and end of `self`, respectively.
+  #
+  # ```
+  # # open ended ranges
+  # s[2..] # => "cde"
+  # s[..2] # => "abc"
+  # ```
+  #
+  # Negative range values are added to `self.size`, thus they are treated as
+  # character indices counting from the end, `-1` designating the last character.
+  #
+  # ```
+  # # negative indices, both ranges are equivalent for `s`
+  # s[1..3]   # => "bcd"
+  # s[-4..-2] # => "bcd"
+  # # Mixing negative and positive indices, both ranges are equivalent for `s`
+  # s[1..-2] # => "bcd"
+  # s[-4..3] # => "bcd"
+  # ```
+  #
+  # Raises `IndexError` if the start index it out of range (`range.begin >
+  # self.size || range.begin < -self.size). If `range.begin == self.size` an
+  # empty string is returned. If `range.begin > range.end`, an empty string is
+  # returned.
+  #
+  # ```
+  # # range.begin > array.size
+  # s[6..10] # raise IndexError
+  # # range.begin == s.size
+  # s[5..10] # => ""
+  # # range.begin > range.end
+  # s[3..1]   # => ""
+  # s[-2..-4] # => ""
+  # s[-2..1]  # => ""
+  # s[3..-4]  # => ""
+  # ```
+  def [](range : Range) : String
     self[*Indexable.range_to_index_and_count(range, size) || raise IndexError.new]
   end
 
-  # Like `#[Range]`, but returns `nil` if the range's start is out of bounds.
+  # Like `#[Range]`, but returns `nil` if `range.begin` is out of range.
```

Comment:
This `#[Range]` syntax is really confusing *and* it is not recognized by the doc generator.

`#[](Range)` would be recognized, though.
Same for `#[](Int, Int)`.

I realize this is pre-existing, but on the other hand, this PR *is* about improving documentation.